File: client/lib/api/v2/schemas/parameters.ts
/**
 * Parameters V2 API Schemas
 * Schema definitions for parameters v2 endpoints with nested items
 */

import { z } from "zod";
import { DepartmentMappingSchema } from "./base";

// ============================================================================
// REQUEST SCHEMAS
// ============================================================================

export const ParametersFiltersSchema = z.object({
  profileId: z.string(),
});

export type ParametersFilters = z.infer<typeof ParametersFiltersSchema>;

// ============================================================================
// RESPONSE SCHEMAS
// ============================================================================

// Sample parameter item for list preview
export const ParameterSampleItemSchema = z.object({
  parameter_item_id: z.string(),
  name: z.string(),
  description: z.string(),
  value: z.string(),
});

export type ParameterSampleItem = z.infer<typeof ParameterSampleItemSchema>;

// Parameter item (normalized)
export const ParameterItemSchema = z.object({
  parameter_id: z.string(),
  name: z.string(),
  description: z.string(),
  numerical: z.boolean(),
  active: z.boolean(),
  department_ids: z.array(z.string()).nullable(), // None = cross-department (all departments)
  num_items: z.number(),
  sample_items: z.array(ParameterSampleItemSchema).default([]),
  can_edit: z.boolean(),
  can_delete: z.boolean(),
  can_duplicate: z.boolean(),
});

export const ParametersListResponseSchema = z.object({
  parameters: z.array(ParameterItemSchema),
});

export type ParametersListResponse = z.infer<
  typeof ParametersListResponseSchema
>;
export type ParameterItem = z.infer<typeof ParameterItemSchema>;

// ============================================================================
// DETAIL SCHEMAS
// ============================================================================

export const ParameterDetailRequestSchema = z.object({
  parameterId: z.string(),
  profileId: z.string(),
});

export type ParameterDetailRequest = z.infer<
  typeof ParameterDetailRequestSchema
>;

// Nested parameter item detail
export const ParameterItemDetailSchema = z.object({
  parameter_item_id: z.string(),
  name: z.string(),
  description: z.string(),
  value: z.string(),
  default_item: z.boolean(),
  department_ids: z.array(z.string()).nullable(), // None = cross-department (all departments)
  can_delete: z.boolean(), // Check if in use
});

export type ParameterItemDetail = z.infer<typeof ParameterItemDetailSchema>;

export const ParameterDetailResponseSchema = z.object({
  // Parameter fields
  name: z.string(),
  description: z.string(),
  numerical: z.boolean(),
  active: z.boolean(),
  department_ids: z.array(z.string()).nullable(),
  valid_department_ids: z.array(z.string()),

  // Nested parameter items
  parameter_items: z.array(ParameterItemDetailSchema),

  // Top-level mappings
  department_mapping: DepartmentMappingSchema,
});

export type ParameterDetailResponse = z.infer<
  typeof ParameterDetailResponseSchema
>;

// Default detail request
export const ParameterDetailDefaultRequestSchema = z.object({
  profileId: z.string(),
});

export type ParameterDetailDefaultRequest = z.infer<
  typeof ParameterDetailDefaultRequestSchema
>;

// ============================================================================
// MUTATION SCHEMAS
// ============================================================================

// Nested parameter item for create/update
export const ParameterItemCreateSchema = z.object({
  name: z.string(),
  description: z.string(),
  value: z.string(),
  default_item: z.boolean(),
  department_ids: z.array(z.string()).nullable().optional(), // Optional for backward compatibility
});

export type ParameterItemCreate = z.infer<typeof ParameterItemCreateSchema>;

// Create request
export const CreateParameterRequestSchema = z.object({
  name: z.string(),
  description: z.string(),
  numerical: z.boolean(),
  active: z.boolean(),
  department_ids: z.array(z.string()).nullable().optional(), // Deprecated - use per-item department_ids
  parameter_items: z.array(ParameterItemCreateSchema),
});

export type CreateParameterRequest = z.infer<
  typeof CreateParameterRequestSchema
>;

export const CreateParameterResponseSchema = z.object({
  success: z.boolean(),
  parameterId: z.string(),
  message: z.string(),
});

export type CreateParameterResponse = z.infer<
  typeof CreateParameterResponseSchema
>;

// Update request
export const UpdateParameterRequestSchema = z.object({
  parameterId: z.string(),
  name: z.string(),
  description: z.string(),
  numerical: z.boolean(),
  active: z.boolean(),
  department_ids: z.array(z.string()).nullable().optional(), // Deprecated - use per-item department_ids
  parameter_items: z.array(ParameterItemCreateSchema),
});

export type UpdateParameterRequest = z.infer<
  typeof UpdateParameterRequestSchema
>;

export const UpdateParameterResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type UpdateParameterResponse = z.infer<
  typeof UpdateParameterResponseSchema
>;

// Duplicate request
export const DuplicateParameterRequestSchema = z.object({
  parameterId: z.string(),
});

export type DuplicateParameterRequest = z.infer<
  typeof DuplicateParameterRequestSchema
>;

export const DuplicateParameterResponseSchema = z.object({
  success: z.boolean(),
  parameterId: z.string(),
  message: z.string(),
});

export type DuplicateParameterResponse = z.infer<
  typeof DuplicateParameterResponseSchema
>;

// Delete request
export const DeleteParameterRequestSchema = z.object({
  parameterId: z.string(),
});

export type DeleteParameterRequest = z.infer<
  typeof DeleteParameterRequestSchema
>;

export const DeleteParameterResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type DeleteParameterResponse = z.infer<
  typeof DeleteParameterResponseSchema
>;

// ============================================================================
// PARAMETER ITEM CREATION (for inline creation from pickers)
// ============================================================================

export const CreateParameterItemRequestSchema = z.object({
  parameterId: z.string(),
  name: z.string(),
  description: z.string(),
  value: z.string(),
  default_item: z.boolean(),
});

export type CreateParameterItemRequest = z.infer<
  typeof CreateParameterItemRequestSchema
>;

export const CreateParameterItemResponseSchema = z.object({
  success: z.boolean(),
  parameterItemId: z.string(),
  message: z.string(),
});

export type CreateParameterItemResponse = z.infer<
  typeof CreateParameterItemResponseSchema
>;

File: server/app/schemas/parameters.py
"""Parameters V2 API schemas with nested items."""

from pydantic import BaseModel

from .base import DepartmentMapping

# ============================================================================
# REQUEST SCHEMAS
# ============================================================================


class ParametersFilters(BaseModel):
    """Filters for parameters list."""

    departmentIds: list[str]
    profileId: str


# ============================================================================
# RESPONSE SCHEMAS
# ============================================================================


class ParameterItem(BaseModel):
    """Parameter item in list response."""

    parameter_id: str
    name: str
    description: str
    numerical: bool
    active: bool
    default_parameter: bool
    num_items: int
    can_edit: bool
    can_delete: bool
    can_duplicate: bool


class ParametersListResponse(BaseModel):
    """Response for parameters list endpoint."""

    parameters: list[ParameterItem]


# ============================================================================
# DETAIL SCHEMAS
# ============================================================================


class ParameterDetailRequest(BaseModel):
    """Request for parameter detail."""

    parameterId: str
    profileId: str


class ParameterItemDetail(BaseModel):
    """Parameter item detail."""

    parameter_item_id: str
    name: str
    description: str
    value: str
    default_item: bool
    can_delete: bool


class ParameterDetailResponse(BaseModel):
    """Response for parameter detail endpoint."""

    # Parameter fields
    name: str
    description: str
    numerical: bool
    active: bool
    default_parameter: bool
    department_id: str
    valid_department_ids: list[str]

    # Nested parameter items
    parameter_items: list[ParameterItemDetail]

    # Top-level mappings
    department_mapping: DepartmentMapping


class ParameterDetailDefaultRequest(BaseModel):
    """Request for default parameter detail."""

    profileId: str


# ============================================================================
# MUTATION SCHEMAS - NESTED STRUCTURES
# ============================================================================


class ParameterItemCreate(BaseModel):
    """Parameter item creation schema."""

    name: str
    description: str
    value: str
    default_item: bool


class CreateParameterRequest(BaseModel):
    """Request to create parameter with nested items."""

    name: str
    description: str
    numerical: bool
    active: bool
    default_parameter: bool
    department_id: str
    parameter_items: list[ParameterItemCreate]


class CreateParameterResponse(BaseModel):
    """Response from create parameter."""

    success: bool
    parameterId: str
    message: str


class UpdateParameterRequest(BaseModel):
    """Request to update parameter with nested items."""

    parameterId: str
    name: str
    description: str
    numerical: bool
    active: bool
    default_parameter: bool
    department_id: str
    parameter_items: list[ParameterItemCreate]


class UpdateParameterResponse(BaseModel):
    """Response from update parameter."""

    success: bool
    message: str


class DuplicateParameterRequest(BaseModel):
    """Request to duplicate parameter."""

    parameterId: str


class DuplicateParameterResponse(BaseModel):
    """Response from duplicate parameter."""

    success: bool
    parameterId: str
    message: str


class DeleteParameterRequest(BaseModel):
    """Request to delete parameter."""

    parameterId: str


class DeleteParameterResponse(BaseModel):
    """Response from delete parameter."""

    success: bool
    message: str


# ============================================================================
# PARAMETER ITEM CREATION (for inline creation from pickers)
# ============================================================================


class CreateParameterItemRequest(BaseModel):
    """Request to create a single parameter item."""

    parameterId: str
    name: str
    description: str
    value: str
    default_item: bool


class CreateParameterItemResponse(BaseModel):
    """Response from create parameter item."""

    success: bool
    parameterItemId: str
    message: str
